File: 20.valid-parentheses.go
package main

/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 */

//  Input: "({[]})"
// Initialize:

// stack = []
// Process each character:

// '(': Add to stack → stack = ['('].
// '{': Add to stack → stack = ['(', '{'].
// '[': Add to stack → stack = ['(', '{', '['].
// ']': Matches '[' → Pop → stack = ['(', '{'].
// '}': Matches '{' → Pop → stack = ['('].
// ')': Matches '(' → Pop → stack = [].
// Final Check:

// stack is empty → Return true.
// @lc code=start
func isValid(s string) bool {
	stack := []rune{}

	bracketMap := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}

	for _, char := range s {
		//查詢char是否在bracketMap中
		//openBracket ->如果是傳入),回傳(
		if openBracket, exits := bracketMap[char]; exits {
			// 如果堆疊為空或者頂部元素不等於對應的開括號
			if len(stack) == 0 || stack[len(stack)-1] != openBracket {
				return false
			}
			//如果最頂部存在對應的key,把char存入stack
			stack = stack[:len(stack)-1]
		} else {
			//如果不存在,把char存入stack
			stack = append(stack, char)
		}
	}
	return len(stack) == 0

	// for _, char := range s {
	// 	if openBracket, exits := bracketMap[char]; exits {
	// 		if len(stack) == 0 || stack[len(stack)-1] != openBracket {
	// 			return false
	// 		}
	// 		stack = stack[:len(stack)-1]
	// 	} else {
	// 		stack = append(stack, char)
	// 	}
	// }
	// return len(stack) == 0

}

// @lc code=end

File: 20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python
#
# [20] Valid Parentheses
#

# Example 1:

# Input: s = "()"

# Output: true

# Example 2:

# Input: s = "()[]{}"

# Output: true


# @lc code=start
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        
        #建立closing bracket
        bracket_map = {')': '(', '}': '{', ']': '['}
        stack = []

        for char in s:
            #如果char有在bracket_map中,表示是closing bracket的
            #把stack彈出
            #存入#是怕stack是空時,會error
            #如果是),},] -> 才會進入比對
            if char in bracket_map:
                top_element = stack.pop() if stack else '#'
                if bracket_map[char] != top_element:
                    return False
            #如果不是closing bracket,把char append到stack中
            else:
                stack.append(char)
        #如果最後stack都是一對,都會pop完,最後為空,會回傳true
        return not stack

        if not stack or stack[-1] != pairs[char]:
            
# @lc code=end

